sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,698 @@
+have the numerical values `8` and `9`.  In addition, the function uses an
+optional feature that corresponds to an aspect with numerical value `8`.
+
+```
+define void @foo() !intel_declared_aspects !1 !intel_used_aspects !2 {}
+!1 = !{i32 8, i32 9}
+!2 = !{i32 8}
+```
+
+
+### Changes to the DPC++ front-end
+
+The front-end of the device compiler is responsible for parsing the
+`[[sycl::requires()]]` and `[[sycl_detail::uses_aspects()]]` attributes and
+transferring the information to the LLVM IR metadata described above according
+to the following rules:
+
+* If the translation unit contains any type definitions that are decorated with
+  `[[sycl_detail::uses_aspects()]]`, the front-end creates an
+  `!intel_types_that_use_aspects` metadata describing the aspects used by all
+  such types.
+
+* If a function is decorated with `[[sycl_detail::uses_aspects()]]`, the
+  front-end adds an `!intel_used_aspects` metadata to the function's definition
+  listing the aspects from that attribute.
+
+* If a function is decorated with `[[sycl::requires()]]`, the front-end adds
+  an `!intel_declared_aspects` metadata to the function's definition listing
+  the aspects from that attribute.
+
+
+### New LLVM IR pass to propagate aspect usage
+
+We add a new IR phase to the device compiler which does the following:
+
+* Creates (or augments) a function's `!intel_used_aspects` metadata with
+  aspects that come from references to types in the
+  `intel_types_that_use_aspects` list.
+
+* Propagates each function's `!intel_used_aspects` metadata up the static call
+  graph so that each function lists the aspects used by that function and by
+  any functions it calls.
+
+* Diagnoses a warning if any function that has `!intel_declared_aspects` uses
+  an aspect not listed in that declared set.
+
+It is important that this IR phase runs before any other optimization phase
+that might eliminate a reference to a type or inline a function call because
+such optimizations will cause us to miss information about aspects that are
+used.  Therefore, it is recommended that this new phase run first, before all
+other IR phases.
+
+Implementing the first bullet point is straightforward.  The implementation can
+scan the IR for each function looking for instructions that reference a type.
+It can then see if that type is in the `!intel_types_that_use_aspects` set; if
+so it adds the type's aspects to the function's `!intel_used_aspects` set.
+While doing this, the implementation must have a special case for the `double`
+type because the front-end does not include that type in the
+`!intel_types_that_use_aspects` set.  If a function references the `double`
+type, the implementation implicitly assumes that the function uses
+`aspect::fp64` and adds that aspect to the function's `!intel_used_aspects`
+set.
+
+The second bullet point requires building the static call graph, but the
+implementation need not scan the instructions in each function.  Instead, it
```

Comment:
We do have to scan instructions in order to build call graph. LLVM doesn't store information about call graph anywhere in `Function`, `Module`, or other entities.

Reply:
I think the call graph can be built during the IR scan that is required for the first bullet.

I wasn't sure how much detail we should add to these design documents.  Do you want me to make a note of that?

Reply:
>I wasn't sure how much detail we should add to these design documents. Do you want me to make a note of that?

No, I think the wording is ok